reps/2023-05-10-actors-batch-remote-api.md

```diff
@@ -0,0 +1,264 @@
+## Summary
+### General Motivation
+**Core Motivation**:
+1. Improve the performance of batch calling ActorTask. The current performance bottleneck of batch invoking ActorTask mainly lies in the duplicate serialization of parameters, repeated puts parameters into object store, and frequent context switching between Python and C++. The Batch remote API will optimize performance for these three aspects.
+
+
+In distributed computing scenarios, such as big data computing、AI training and inference, there is a need to send a large number of RPC requests(ActorTask) to multiple Actors in batches.  
+For example, in a typical star-topology architecture with one master Actor and 400 worker Actors, the same computational request needs to be sent to 400 worker Actors. 
+However, the computing tasks of each Worker Actor are very short. In such a scenario, the performance requirements for executing batch actor task remote of a large number of Actor are very high.  
+Therefore, for the scenario of batch calling Actor tasks, I want to add a new optimization API, batch_remote(), to improve the performance of batch submission of Actor Task calls.   
+After my own performance testing and comparison, this API interface can improve performance by 2 ~ 100+ times.
+
+Current situation of batch calling actor tasks:
+```
+actors = [WorkerActor.remote() for _ in range(400)]
+
+# This loop's repeated invocation actually wastes a lot of performance.
+for actor in actors:
+  actor.compute.remote(args)
+```
+
+Using the new Batch Remote API:
+```
+actors = [WorkerActor.remote() for _ in range(400)]
+
+# Calling it only once can greatly improve performance.
+batch_remote_handle = ray.experimental.batch_remote(actors)
+batch_remote_handle.compute.remote(args)
+```
+
+The current performance bottleneck of batch invoking ActorTask mainly lies in the duplicate serialization of parameters, repeated puts parameters into object store, and frequent context switching between Python and C++.  
+The Batch Remote API can reduce the following performance costs(The N is the number of Actors):  
+1. Reduce (N-1) times of parameter serialization performance costs.
+2. Reduce (N-1) times of putting parameter into object store performance costs for scenarios with large parameters.
+3. Reduce (N-1) times of context switching between Python and C++ and repeated parameter verification performance costs.
+
+
+### Should this change be within `ray` or outside?
+This requires adding a new interface in Ray Core.  
+The initial consideration is to add it to the ray.experimental module.
+
+## Stewardship
+### Required Reviewers
+...
+
+### Shepherd of the Proposal (should be a senior committer)
+...
+
+## Design and Architecture
+### API
+Use case  
+Plan 1
+```
+batch_remote_handle = ray.experimental.batch_remote(actors)
+batch_remote_handle.compute.remote(args)
```

Comment:
what happens if one of the actor failed? (i.e. killed or terminated?)

Reply:
Failure & Exception Scenario.
**1. Exceptions occurred during parameter validation or preprocessing before batch submission of ActorTasks.** 
Since these exceptions occur before the process of submitting ActorTasks, they can be handled by directly throwing specific error exceptions as current situation.

**2. Some actors throw exceptions during the process of batch submitting ActorTasks.** 
When traversing and submitting ActorTasks in a loop, if one of the Actors throws an exception during submission, the subsequent ActorTasks will be terminated immediately, and the exception will be throwed to user.

Reason:
Submitting ActorTask is normally done without any exceptions being thrown. If an error does occur, it is likely due to issues with the code and will require modifications.
The exception behavior of this plan is the same as the current foreach remote.

![Uploading image.png…]()